### PR TITLE
Detect the proper start and end position of a line.

### LINKED ISSUE #1222: Phpstan plugin's error underline ignores whitespaces
When underlining the offending line of code, the Phpstan plugin seems to ignore the leading whitespace but doesn't adjust the length for it:

image

### PR DESCRIPTION
**Pull request by BladeMF**
Fixes https://github.com/phpactor/phpactor/issues/1222 and https://github.com/phpactor/language-server-phpstan-extension/issues/4.

### CODE REVIEW

File: lib/Model/DiagnosticsParser.php [hunk 1]
```diff
@@ -17,21 +17,30 @@ public function parse(string $jsonString): array
...
+            }
+                 
             foreach ($fileDiagnostics['messages'] as $message) {
                 $lineNo = (int)$message['line'] - 1;
                 $lineNo = (int)$lineNo > 0 ? $lineNo : 0;
-
+                $line = $lines[$lineNo] ?? "";
+                 
+                $matches = [];
+                $offset = (\preg_match("/^\\s+/", $line, $matches) === 1) ? mb_strlen($matches[0]) : 0;
```
Comment (dantleech) [inline lib/Model/DiagnosticsParser.php]: I created a utility for this: https://github.com/phpactor/text-document/blob/master/lib/Util/LineColRangeForLine.php might be better to use that

Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: `(new LineColRangeForLine())->rangeFromLine($text, $lineNo)`

Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: For some reason I don't have this in my version of text-document.

Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: yes you probably need to bump to the next patch version

Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: `"phpactor/text-document": "^1.2.4"`

Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: `"phpactor/language-server-extension": "^0.6.1",` this is bringing text-document, but it doesn't get me the latest.

Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: (I suck at this)

Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: you need to explicitly require it in composer.json :)

Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: Yes, that worked. Now the util didn't do what I needed, debugging now.

Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: Why does this add 1 to start?
```
       return new LineColRange(
            new LineCol($lineNo, $start + 1),
            new LineCol($lineNo, $end)
        );
```
and not to end, that is.

File: lib/Model/DiagnosticsParser.php [hunk 2]
```diff
@@ -18,20 +19,30 @@ public function parse(string $jsonString): array
         $decoded = $this->decodeJson($jsonString);
         $diagnostics = [];
 
-        foreach ($decoded['files'] ?? [] as $fileDiagnostics) {
+        foreach ($decoded['files'] ?? [] as $uri => $fileDiagnostics) {
+            $text = \file_get_contents($uri);
```
Comment (dantleech) [inline lib/Model/DiagnosticsParser.php]: maybe something like:
```
if (false === $text = @file_get_contents($uri)) {
   continue;
}
```

File: lib/Model/DiagnosticsParser.php [hunk 3]
```diff
@@ -18,20 +19,30 @@ public function parse(string $jsonString): array
...
+            
             foreach ($fileDiagnostics['messages'] as $message) {
                 $lineNo = (int)$message['line'] - 1;
                 $lineNo = (int)$lineNo > 0 ? $lineNo : 0;
-
+                
+                $start = 0;
+                $end = 100;
+                if($text !== false) {
```
Comment (dantleech) [inline lib/Model/DiagnosticsParser.php]: no need for this conditional or preceding assignments if we exit early above

File: lib/Model/DiagnosticsParser.php (hunk 1 above)
Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: good question

Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: not critical here i guess and can be fixed independently if it is wrong

Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: As far as I understand it, it assumes that the line is passed as 1-based index, so it returns start and end as 1 based indices as well. That means probably that end should be `-1` as well. I leave that to you. In this code if I skip to substract 1 from end, the worst case scanerio is that the end will be 1 more than the end of line, which VS will truncate.

Comment (dantleech) [reply lib/Model/DiagnosticsParser.php]: Just to make things interesting the LSP Position is 0 based.

**Comment by dantleech** [comment]
Cheers @BladeMF

File: lib/Model/DiagnosticsParser.php (hunk 1 above)
Comment (BladeMF) [reply lib/Model/DiagnosticsParser.php]: That is why I removed 1 :-)

**Pull request merged**